substrate/frame/election-provider-multi-block/src/signed/tests.rs

```diff
@@ -224,16 +231,364 @@ mod calls {
+
+			// expected base deposit with 0 submissions in the queue.
+			let base_deposit = <Runtime as Config>::DepositBase::convert(0);
+			let page_deposit = <Runtime as Config>::DepositPerPage::get();
+			assert!(base_deposit != 0 && page_deposit != 0 && base_deposit != page_deposit);
+
+			let account_id = 99;
+
+			// account_id has 100 free balance and 0 held balance for elections.
+			// As configured in
+			// pallet-epm-mb/substrate/frame/election-provider-multi-block/src/mock/mod.rs ->
+			// pallet_balances::GenesisConfig::<T>
+			assert_eq!(balances(account_id), (100, 0));
+
+			assert_ok!(SignedPallet::register(
+				RuntimeOrigin::signed(account_id),
+				Default::default()
+			));
+
+			// free balance and held deposit updated as expected.
+			assert_eq!(balances(account_id), (100 - base_deposit, base_deposit));
+
+			// submit page
+			assert_ok!(SignedPallet::submit_page(
+				RuntimeOrigin::signed(account_id),
+				0,
+				Some(Default::default())
+			));
+
+			// free balance and held deposit updated as expected
+			assert_eq!(
+				balances(account_id),
+				(100 - base_deposit - page_deposit, base_deposit + page_deposit)
+			);
+
+			let bailing_account_id = 91;
+
+			// bailing_account_id has 100 free balance and 0 held balance for elections.
+			// As configured in
+			// pallet-epm-mb/substrate/frame/election-provider-multi-block/src/mock/mod.rs ->
+			// pallet_balances::GenesisConfig::<T>
+			assert_eq!(balances(bailing_account_id), (100, 0));
+
+			// account 1 submitted nothing, so bail should have no effect and return error
+			assert_err!(
+				SignedPallet::bail(RuntimeOrigin::signed(bailing_account_id)),
+				SubmissionNotRegistered::<Runtime>
+			);
+
+			// balance of account with a submission as before
+			assert_eq!(
+				balances(account_id),
+				(100 - base_deposit - page_deposit, base_deposit + page_deposit)
+			);
+
+			// bailing account balance unchanged
+			assert_eq!(balances(bailing_account_id), (100, 0));
+		})
+	}
+
+	#[test]
+	fn force_clear_submission_fails_if_called_by_account_none() {
+		ExtBuilder::default().build_and_execute(|| {
+			assert_err!(
+				SignedPallet::force_clear_submission(RuntimeOrigin::none(), 0, 99),
+				BadOrigin
+			);
+		})
+	}
+
+	#[test]
+	fn force_clear_submission_fails_if_called_in_phase_other_than_off() {
+		ExtBuilder::default().build_and_execute(|| {
+			let some_bn = 0;
+			let some_page_index = 0;
+
+			let phases = vec![
+				Phase::Off,
+				Phase::Snapshot(some_page_index),
+				Phase::SignedValidation(some_bn),
+				Phase::Unsigned(some_bn),
+				Phase::Export(some_bn),
+				Phase::Emergency,
+			];
+
+			let account_id = 99;
+			for phase in phases {
+				roll_to_phase(phase);
+
+				assert_err!(
+					SignedPallet::force_clear_submission(RuntimeOrigin::root(), 0, account_id),
+					NotAcceptingSubmissions::<Runtime>,
+				);
+			}
+		})
+	}
+
+	#[test]
+	fn force_clear_submission_fails_if_submitter_done_no_submissions_at_all() {
```

Comment:
Is there a specific reason why you're suggesting the approach of having few tests with lots of checks inside each one?
My preferred approach is exactly the opposite one, so I'd like to understand your point of view.

Reply:
The reasons why I'd merge these small tests cases are:
1. Simplify - we can re-use setup code across tests cases;
2. Context - the tests are basically testing the same unit (the force_clear) behaviour
3. Follow-along/readability - it's easier to undestand all the invariants of `force_clear` etc if they are all in the same test, we can use comments to explain what is being tested and why errors happen, etc
4. Low verbosity - especially due to 1.

See, for example, this test case in staking: https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/staking/src/tests.rs#L1283

We test within a test all the flow of `Call::unbound`. We could split it in different tests but the setup/boilerplate code would be repeated all over the individual tests. Also through that test you can read and understand the whole unbound flow, requirements, invariants, etc

Reply:
https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/staking/src/tests.rs#L1495 being another good example, where `assert_noop!` is used too to ensure that errors do not change the storage root (and state).

Reply:
I think our dispute boils down to the level at which we see these tests. 
You (and maybe it's the way in the whole repo, I can't know yet) advocate for a higher level tests, where a single test function would run a validation flow through the tested function with some failure checks alongside. This makes the test functionality test or flow test or smth like that. So it reads like a story.
My initial approach is in general to test the smallest possible units of functionality first then go to flow tests.
So each unit test function is called `fn_results_with_x_when_z` where conditions and expected results are clear in the name of the test. This way expectations for a given function are set in the unit tests.
Then I go for writing functional/flow tests and I do them as you would, just with less failure checks.

As to the verbosity and boilerplate, it can be often packed in setup functions (to avoid duplication) but it's rather good to be explicit in test setup so as to describe the initial conditions.

So to sum up, I think we actually align on the way to write tests, it's just that I went lower in the test pyramid, which I would argue is a good thing and to improve readability we could arrange the tests differently.